Prevent special chars in case search form

northwestwitch opened the pull request:
Fix #3513 

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the paxa software installed on `hasta`:
 - Log in into hasta: `ssh <USER.NAME>@hasta.scilifelab.se`
 - Activate the staging environment with the command `us` (Use Stage)
 - Run the command `paxa` and follow the instructions. Make sure to specify scout-stage as the resource you request and the server cg-vm1 as server.
1. Log out from the hasta server.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, log out from `cg-vm1` and log in again in the `hasta` server, repeat the `hasta` and `paxa` procedure, which will release the allocated resource (scout-stage) to be used for testing by other users.
</details>


**How to test**:
1. Locally or on cg-vm1
2. Go to an institute and try to type any of the following chars in the case search form: ^<>?!=\\/
3. The form should not validate and display an error
4. Try to reproduce the error from #3513 --> assignee's email search with `\    -` (doesn't matter how many spaces between the 2 characters)
5. The form should display error on submit, while on main branch the page should crash

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

dnil commented on scout/server/blueprints/institutes/templates/overview/cases.html:
```diff
@@ -28,8 +28,8 @@
-        {{ form.search_term.label(class="control-label") }}
-        {{ form.search_term(class="form-control", placeholder='example:18201') }}
+          <label class="control-label" for="search_term">Search cases</label>
+          <input class="form-control" id="search_term" name="search_term" placeholder="example:18201" type="text" pattern="[^<>?!=\\/]*" title="Characters ^<>?!=\/ are not allowed">
```
```suggestion
          <input class="form-control" id="search_term" name="search_term" placeholder="example:18201" type="text" pattern="[^<>?!=\\/]*" data-bs-toggle="tooltip" title="Characters ^<>?!=\/ are not allowed">
```

northwestwitch replied:
I tried locally and your suggestion doesn't display a tooltip. Why a tooltip BTW? This is form validation works fine already:

<img width="442" alt="image" src="https://user-images.githubusercontent.com/28093618/184872612-1424be01-3fec-4f9e-a486-682363766225.png">

northwestwitch replied:
Ahh OK, title is actually used by the HTML5 validation to display the error!

dnil replied:
It seems to be rather variable across user agents; see https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern. Oh, one thing, the caret should not be in the title list - it is the NOT for the pattern match operation. 😊

dnil reviewed (approved):
I can reproduce the issue. This is a nice way to solve it, as it directly informs the user, but it doesn't quite go to the root of it.
It would also feel satisfying to fix the issue at
https://github.com/Clinical-Genomics/scout/blob/9b17661bfdf497c6606cdb2c18a89e142594eb90/scout/adapter/mongo/case.py#L234.

At any rate, as discussed, lets not forget to do the same on the dashboard! 😸

northwestwitch commented:
> I can reproduce the issue. This is a nice way to solve it, as it directly informs the user, but it doesn't quite go to the root of it. It would also feel satisfying to fix the issue at
> 
> https://github.com/Clinical-Genomics/scout/blob/9b17661bfdf497c6606cdb2c18a89e142594eb90/scout/adapter/mongo/case.py#L234
> 
> .
> At any rate, as discussed, lets not forget to do the same on the dashboard! 😸

Fixed by introducing a re.escape to any search term passed to the query. Fixed also the form field in the dashboard page

dnil commented on scout/adapter/mongo/case.py:
```diff
@@ -233,7 +234,9 @@ def _populate_name_query(self, query, name_query, owner=None, collaborator=None)
             user_query = {
-                "$and": [{"name": {"$regex": term, "$options": "i"}} for term in query_terms]
+                "$and": [
+                    {"name": {"$regex": re.escape(term), "$options": "i"}} for term in query_terms
```
👍

dnil commented on scout/adapter/mongo/case.py:
```diff
@@ -156,7 +157,7 @@ def _populate_name_query(self, query, name_query, owner=None, collaborator=None)
         order = None
         query_field = name_query.split(":")[0]  # example:status
-        query_term = name_query[name_query.index(":") + 1 :].strip()
+        query_term = re.escape(name_query[name_query.index(":") + 1 :].strip())
```
👍

dnil commented:
For reference, a loooong hover with Safari will by default show the `title` as if a tooltip, but it does not make it part of the error message.
